src/sdam/description/topology/server_selection/mod.rs

```diff
@@ -1,25 +1,91 @@
 #[cfg(test)]
 mod test;
 
-use std::{fmt, time::Duration};
+use std::{collections::HashMap, fmt, ops::Deref, sync::Arc, time::Duration};
 
-use rand::seq::IteratorRandom;
+use rand::{
+    rngs::SmallRng,
+    seq::{IteratorRandom, SliceRandom},
+    SeedableRng,
+};
 
 use super::TopologyDescription;
 use crate::{
     error::{ErrorKind, Result},
+    options::StreamAddress,
     sdam::{
         description::{
             server::{ServerDescription, ServerType},
             topology::TopologyType,
         },
         public::ServerInfo,
+        Server,
     },
     selection_criteria::{ReadPreference, SelectionCriteria, TagSet},
 };
 
 const DEFAULT_LOCAL_THRESHOLD: Duration = Duration::from_millis(15);
 
+/// Struct encapsulating a selected server that handles the opcount accounting.
+#[derive(Debug)]
+pub(crate) struct SelectedServer {
+    server: Arc<Server>,
+}
+
+impl SelectedServer {
+    fn new(server: Arc<Server>) -> Self {
+        server.increment_operation_count();
+        Self { server }
+    }
+}
+
+impl Deref for SelectedServer {
+    type Target = Server;
+
+    fn deref(&self) -> &Server {
+        self.server.deref()
+    }
+}
+
+impl Drop for SelectedServer {
+    fn drop(&mut self) {
+        self.server.decrement_operation_count();
```

Comment:
I'm not sure this is correct; we generally stop using the server after we've checked out a connection, and the compiler is free to drop the server at any point after the last usage, which could be before the current scope ends.

Reply:
We continue using the checked out server until the operation we execute on it is done (see [here](https://github.com/patrickfreed/mongo-rust-driver/blob/RUST-575/health-rebased/src/client/executor.rs#L115) for example). Additionally, I don't believe the compiler is allowed to arbitrarily drop things when they're not being used. They have to be dropped at the end of a "drop scope" which is explcitly defined ([docs](https://doc.rust-lang.org/reference/destructors.html#drop-scopes)). If this wasn't the case, lock guards would be pretty cumbersome to use.

Reply:
Interesting; I had thought they dropped that guarantee when they switched to non-lexical lifetimes, but I must have been mistaken.